packages/relay-compiler/generic/codegen/RelayCodegenWatcher.js

```diff
@@ -35,23 +44,52 @@ type WatchmanChanges = {
   files?: Array<WatchmanChange>,
 };
 
+function buildWatchExpression(options: FilterOptions): WatchmanExpression {
+  return [
+    'allof',
+    ['type', 'f'],
+    ['anyof', ...options.extensions.map(ext => ['suffix', ext])],
+    [
+      'anyof',
+      ...options.include.map(include => ['match', include, 'wholename']),
+    ],
+    ...options.exclude.map(exclude => ['not', ['match', exclude, 'wholename']]),
+  ];
+}
+
 async function queryFiles(
-  baseDir: string,
-  expression: WatchmanExpression,
+  filterOptions: FilterOptions,
   filter: FileFilter,
 ): Promise<Set<File>> {
-  const client = new RelayWatchmanClient();
-  const [watchResp, fields] = await Promise.all([
-    client.watchProject(baseDir),
-    getFields(client),
-  ]);
-  const resp = await client.command('query', watchResp.root, {
-    expression,
-    fields: fields,
-    relative_root: watchResp.relativePath,
-  });
-  client.end();
-  return updateFiles(new Set(), baseDir, filter, resp.files);
+  const {baseDir, include, extensions, exclude} = filterOptions;
+  if (
```

Comment:
Let's refactor this into two functions since this is getting a bit complicated. How about `queryFilesWithWatchman` and `queryFilesWithGlob` and you can call them from this condition?